### ISSUE TITLE
Slow open RLMRealm on Realm 0.93.2 and greater (0.95.0 included)

### ISSUE THREAD

**Issue by javialgaba** [realm/realm-cocoa#2510]
I have found a regression in **realm-cocoa**. Version *0.93.1* and before load the first `RLMRealm` instance of a .realm file much much faster than newer versions. I have attached a sample code where you can compare results.

I am developing an app with lots of different Realms, and this bug has forced me to downgrade to Realm 0.93.1. Before opening this issue, I have tested different realm-cocoa versions, and I can confirm that this problem start to arise at version **0.93.2**.

Anyway, thank you very much for your amazing work. Realm is really great :)

```
- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Load instance after a couple of seconds to get more time accuracy
    dispatch_queue_t queue = dispatch_get_main_queue();
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 2.0 * NSEC_PER_SEC), queue, ^{
        NSTimeInterval startTime = CACurrentMediaTime();
        [RLMRealm defaultRealm]; // Load a Realm instance
        NSLog(@"realm loaded (s): %f", CACurrentMediaTime()-startTime);
    });
}
```

Here is a simple benchmark I have done comparing *0.93.1* and *0.95.0* (time in seconds)

| Realm 0.93.1 | Realm 0.95.0  |
| ------------------ | ------------------ |
| 0,025706       | 0,598671        |
| 0,015229       | 0,558317        |
| 0,015650       | 0,586977        |
| 0,016144       | 0,788107        |
| 0,015895       | 0,545924        |
| 0,016177       | 0,562112        |
| 0,016205       | 0,565462        |
| 0,016735       | 0,514518        |
| 0,017433       | 0,578577        |
| 0,015473       | 0,564286        |

**Average**

| Realm 0.93.1 | Realm 0.95.0 |
| ------------------ | ------------------ |
| 0,0170647     | 0,5862951      |

We are seeing here that v0.93.1 is **3435,72% faster** on average loading the first realm instance

**Comment by bmunkholm** [comment]
Wauu - that does not look good! We will look into that for sure.
Thanks for the feedback!

**Comment by jpsim** [comment]
Thanks for reporting this @javialgaba. We'll look into this, but before we dive in, I want to share my preliminary thoughts.

In 0.93.2, we introduced #2077, which we knew would cause performance regressions in some cases:

> Fix a case where data loss would occur if a device was hard-powered-off shortly after a write transaction was committed which had to expand the Realm file.

The core PR https://github.com/realm/realm-core/pull/875 has a bit more information:

> Fixed durability issue in case of power / system failures on Apple platforms. We now use a stronger synchronization (`fcntl(fd, F_FULLFSYNC)`) to stable storage when the file is extended.

We do this to work around issues in Apple platforms' file system that don't fully flush out data to disk in a permanent state in cases where the device powers down shortly after a write. This caused data corruption in some rare cases, and so preventing that was worth introducing a performance regression when expanding the file (which only happens in certain intervals).

So to recap, this performance regression may be the result of this intentional change. But we'll still look into it further to see if there are any other factors contributing to the longer open times for your file.

**Comment by tgoyne** [comment]
To clarify, this is when the Realm file doesn't already exist, as opposed to the first time opening an existing Realm file? I wrote the following simple performance test for this and ran it on both 0.93.1 and 0.93.2:

```objc
- (void)testRealmFileCreation {
    __block int measurement = 0;
    const int iterations = 50;
    NSString *path = [[[RLMRealm defaultRealmPath] stringByDeletingLastPathComponent] stringByAppendingPathComponent:@"dir"];
    [NSFileManager.defaultManager removeItemAtPath:path error:nil];
    [NSFileManager.defaultManager createDirectoryAtPath:path withIntermediateDirectories:YES attributes:nil error:nil];
    [self measureBlock:^{
        for (int i = 0; i < iterations; ++i) {
            @autoreleasepool {
                [RLMRealm realmWithPath:[path stringByAppendingPathComponent:@(measurement * iterations + i).stringValue]];
            }
        }
        ++measurement;
    }];
}
```

This gave an average of 2.819 second with 0.93.1 and 7.358 seconds with 0.93.2, which is an unpleasant hit (creating the file is pretty much the worst possible case for the change jpsim mentioned), but a significantly smaller difference than what you're seeing.

One possibility is that since `F_FULLFSYNC` waits for all of the drive's buffers to be flushed, you're seeing significant delays from it waiting for writes from outside your app to complete, which wouldn't show in in a small artificial test. If that's the case, creating a second Realm file at a different path will take significantly less time than creating the first one did.

**Comment by javialgaba** [comment]
First of all, thank you very much for you quick response ;)

@jpsim #2077 was also my main thought. But I still find this incredible slow. 

I have run the test @tgoyne posted and I dont get the same results. Not even close. I know I'm running the tests on an iMac (Late 2012) without SSD, but I'm getting values much much worse. In fact, I've tweaked the test to do only 10 runs instead of 50 because it took really long to execute on 0.93.2.

I share my results with you.

```
- (void)testRealmFirstCreation {
    NSString *path = [[[RLMRealm defaultRealmPath] stringByDeletingLastPathComponent] stringByAppendingPathComponent:@"dir"];
    [NSFileManager.defaultManager removeItemAtPath:path error:nil];
    [NSFileManager.defaultManager createDirectoryAtPath:path withIntermediateDirectories:YES attributes:nil error:nil];
    const NSInteger maxIterations = 10;
    [self measureBlock:^{
        for (NSInteger index = 0; index < maxIterations; index++) {
            @autoreleasepool {
                NSString *realmAtPath = [[path stringByAppendingPathComponent:@(index).stringValue] stringByAppendingString:@".realm"];
                [RLMRealm realmWithPath:realmAtPath];
            }
        }
    }];
}
```

**Realm 0.93.1**

| measureBlock | full test |
| ------------ | --------- |
| 0,015        | 0,567     |
| 0,016        | 0,425     |
| 0,017        | 0,438     |
| 0,014        | 0,406     |
| 0,014        | 0,403     |

**Realm 0.93.2**

| measureBlock | full test |
| ------------ | --------- |
| 4,082        | 41,089    |
| 3,845        | 38,717    |
| 4,007        | 40,332    |
| 3,955        | 39,811    |
| 3,979        | 40,049    |

Thank you very much for your help.

**Comment by tgoyne** [comment]
The delta being much larger on OS X would make sense (I tested on an iOS device) since there's a lot more i/o going on in the background and bigger caches to flush. I'll double-check that we're not fullsyncing more than we need to, but I suspect that the only resolution will be to add a lower durability mode.

**Comment by jpsim** [comment]
I just created #2603 to track adding the ability to run Realm without some durability guarantees in the cases where performance is more important than avoiding corrupted files.
